src/observed/abstract.jl

```
@@ -8,3 +8,142 @@ Rows are samples, columns are observed variables.
+# reorder the columns of the data to match the speciation.
+# If no variable names are provided at all, generates the symbols in the form
+# Symbol(observed_var_prefix, i) for i=1:nobserved_vars.
+function prepare_data(
+    data::Union{AbstractDataFrame, AbstractMatrix, Nothing},
+    observed_vars::Union{AbstractVector, Nothing},
+    spec::Union{SemSpecification, Nothing},
+    nobserved_vars::Union{Integer, Nothing} = nothing;
+    observed_var_prefix::Union{Symbol, AbstractString},
+)
+    obs_vars = nothing
+    obs_vars_perm = nothing
+    if !isnothing(observed_vars)
+        obs_vars = Symbol.(observed_vars)
+        if !isnothing(spec)
+            obs_vars_spec = SEM.observed_vars(spec)
+            try
+                obs_vars_perm = source_to_dest_perm(
+                    obs_vars,
+                    obs_vars_spec,
+                    one_to_one = false,
+                    entities = "observed_vars",
+                )
+            catch err
+                if isa(err, KeyError)
+                    throw(
+                        ArgumentError(
+                            "observed_var \"$(err.key)\" from SEM specification is not listed in observed_vars argument",
+                        ),
+                    )
+                else
+                    rethrow(err)
+                end
+            end
+            # ignore trivial reorder
+            if obs_vars_perm == eachindex(obs_vars)
+                obs_vars_perm = nothing
+            end
+        end
+    elseif !isnothing(spec)
+        obs_vars = SEM.observed_vars(spec)
+    end
+    # observed vars in the order that matches the specification
+    obs_vars_reordered = isnothing(obs_vars_perm) ? obs_vars : obs_vars[obs_vars_perm]
+
+    # subset the data, check that obs_vars matches data or guess the obs_vars
+    if data isa AbstractDataFrame
+        if !isnothing(obs_vars_reordered) # subset/reorder columns

```

Should we throw an error if someone does something like

```
model = Sem(
    specification = par_table,
    data = example_data("political_democracy"),
    observed_vars = observed_vars[randperm(11)],
    loss = SemML,
    meanstructure = false,
)
```

i.e. pass a `DataFrame` and `observed_vars` with a different order of variables - bcs at the moment, `observed_vars` is just ignored in this case. I'm not sure there are any cases where it makes sense to pass both a `DataFrame` and `observed_vars`, since the data frame has names that should be used to reorder according to the `specification` object.

---

I have added a commit that shows a warning in such cases (DataFrame, mismatch in `observed_vars` and `observed_vars(specification)`).
I think we write in the docstrings that the specification has a precedence in defining the final order of the variables, so it is not necessary to throw an error.
If the user thinks that *observed_vars* kw could be used to rename the *DataFrame* columns, that warning should clarify the confusion.

---

Great, thanks.